gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeUtils.java

```diff
@@ -229,4 +229,30 @@ public static boolean areFiltersOnlyOnCurrentDataSource(
         .reduce(Sets::intersection)
         .orElse(Collections.emptySet());
   }
+
+  /**
+   * Removes duplicate selection attributes from other sources using {@param source} as the pivot
+   * source
+   */
+  public static void removeDuplicateSelectionAttributes(
+      ExecutionContext executionContext, String source) {
+    Map<String, Set<String>> sourceToSelectionAttributeMap =
```

Comment:
nit: Move the copy after the if below. Can avoid the unnecessary copy in early return cases.